### PR TITLE
ARROW-10196: [C++] Add Future::DeferNotOk

### PR DESCRIPTION
**Pull request by bkietz**

### CODE REVIEW

File: cpp/src/arrow/util/future.h [hunk 1]
```diff
@@ -154,9 +154,8 @@ class FutureStorage : public FutureStorageBase {
 
   Status status() const { return result_.status(); }
 
-  template <typename U>
-  void MarkFinished(U&& value) {
-    result_ = std::forward<U>(value);
+  void MarkFinished(Result<T> result) {
```
Comment (pitrou) [inline cpp/src/arrow/util/future.h:157]: Does this implicitly convert from any T-convertible U?

File: cpp/src/arrow/util/future.h [hunk 2]
```diff
@@ -358,6 +357,13 @@ class Future {
     return fut;
   }
 
+  static Future DeferNotOk(Result<Future> maybe_future) {
```
Comment (pitrou) [inline cpp/src/arrow/util/future.h:360]: Can you add a docstring?

**Review by pitrou — COMMENTED** [review]
+1 on the principle, just two comments.

File: cpp/src/arrow/util/future.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/util/future.h:360]: Will do

File: cpp/src/arrow/util/future.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/util/future.h:157]: Yes, because `Result<T>` already has that implicit constructor

**Review by pitrou — APPROVED** [review]
+1, will merge if green

**Comment by pitrou** [comment]
Flight test failure on AppVeyor is unrelated.

**Pull request closed without merging**